### DISCUSSION TITLE
File handling within FP: How to pull files out of map?

### DISCUSSION THREAD

**Discussion by theelectricsoul** [freeplane/freeplane#1222] [Help]
Hello everybody :)

**Dragging files on to map ✅**
It is possible to drag a file on to a node and FP creates either a copy or a reference to that file. Nice.

**Pulling files out of map ❌**
At some point I need to pull files out. I have not found way to do so. I need to figure out the path and the name of the file and copy it by hand. Is there a convenient way, besides scripting?

**Comment by macmarrum** [comment]
As far as I understand, you are talking about a node with a link to a file. When hovering mouse coursor over the link icon, the status bar displays the file path including its name.

![openDirectoryOfLinkedFile](https://github.com/freeplane/freeplane/assets/7197131/be0be6d7-35e7-4a32-95d1-3780ff963c3a)


To access such files, I use the script [openDirectoryOfLinkedFile.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/scripts/openDirectoryOfLinkedFile.groovy) and work with the file in the File Manager.

_Note: Change or remove `="/menu_bar/Mac2"` in the first line to have it appear in a different menu location_

**Reply by theelectricsoul** [reply]
Thank you for your answer.

I was not precise... sry. I know they I can see the filepath in FP. But I am looking for a way to handle files a la Drag and Drop or anything that makes pulling files out easy. Even better would be, if the files that are copied into the map(not just referenced) would be delete, if they are no present on the map.

I assume.... this can be a basis for a feature request :)

**Reply by macmarrum** [reply]
When you say "copied into the map (not just referenced)", what do you mean? I cannot recall such functionality.

What would be your expectation of this "pulling a file" in the mindmap? Would the corresponding node be deleted?

**Reply by theelectricsoul** [reply]
Have a look :)
I talk about copying files into the map, deleting and handling them.


https://github.com/freeplane/freeplane/assets/10260832/fa2c8010-d56d-4aea-b30a-44e15a34e3c0

**Reply by abc16361** [reply]
I use this simple script to copy node links (one or multiple selected nodes) to clipboard. Do you think it helps?

```
//@ExecutionModes({ON_SINGLE_NODE})

def getNodes = c.selecteds
def links = ""
getNodes.each{ nd ->
	def lnk = nd.link.text
	links =  links + '\n' + lnk
}	
textUtils.copyToClipboard("${links}")
```

**Reply by theelectricsoul** [reply]
> I use this simple script to copy node links (one or multiple selected nodes) to clipboard. Do you think it helps?
> 
> ```
> //@ExecutionModes({ON_SINGLE_NODE})
> 
> def getNodes = c.selecteds
> def links = ""
> getNodes.each{ nd ->
> 	def lnk = nd.link.text
> 	links =  links + '\n' + lnk
> }	
> textUtils.copyToClipboard("${links}")
> ```

Is definitely helpful. :) Thank you

**Reply by abc16361** [reply]
I modified it a bit. It does not copy 'null' link if a node without link is mistakenly selected and removes the first blank line on the result. You can test it to see which is better.

```
//@ExecutionModes({ON_SINGLE_NODE})

def getNodes = c.selecteds
def links = ""
getNodes.each{ nd ->
	def lnk = nd.link.text
	if ( lnk != null ) {
		links =  links + '\n' + lnk
	}
}	
links = links.trim()
textUtils.copyToClipboard("${links}")
```

**Reply by theelectricsoul** [reply]
The script is helpful if you pull links one by one. Because usually you cannot throw multiple links into email software or similar. ...well in my experience

**Reply by abc16361** [reply]
The script works on One or Multiple selected nodes. If you need to get links one by one, just select one node and run the script each time.

**Reply by theelectricsoul** [reply]
@macmarrum 
It did not work on a node, as it returned a bad uri. 
![image](https://github.com/freeplane/freeplane/assets/10260832/5140596d-62e4-429d-8256-1f772c529ba7)



It changed it, and now it opens the files directly. Also I do not know anything about java/groovy
```
// @ExecutionModes({ON_SELECTED_NODE="/menu_bar/Mac2"})
// based on OpenCurrentMapDirAction.java
import org.freeplane.core.util.Hyperlink
import org.freeplane.features.link.LinkController
import org.freeplane.plugin.script.proxy.ScriptUtils

def node = ScriptUtils.node()
def c = ScriptUtils.c()

if (node.link.file) {
//    def uri = node.link.file.parentFile.toURI() //OLD01
    def uri = node.link.uri  //NEW01
    LinkController.controller.loadHyperlink(new Hyperlink(uri))
} else {
    c.statusInfo = 'the node is missing a link of type "file"'
}
```

**Reply by theelectricsoul** [reply]
@abc16361 
This script results in a URL looking like this. What `../../` stand for ? 
`../../K:/Futi/Schulden/2023.02.18%20-%20Futi%20Schulden%20Pair%20Finance.pdf`

**Reply by abc16361** [reply]
I think most likely you have set the hyperlink type in a relative link format that brings out that pattern. I use absolute link, so it would add ' file:/ ' in front of a local path on Freeplane.
<img width="597" alt="hyperlinkType" src="https://github.com/freeplane/freeplane/assets/101362122/ac41dbf9-31db-41e9-8fce-f2fb683c86f7">

**Comment by EdoFro** [comment]
Here is a script to clean the '_files' folder associated to the map. It deletes all files that are no longer linked to the map (not present as an externalObject image or as a node's link to a file)

Attention: 
It doesn't look into the attributes or the rich text contents to see if the files may be linked from there. So if you linked a file that way, it will delete it, because it won't know it is linked to the map.

here is the script:

```groovy
import groovy.io.FileType
import groovy.io.FileVisitResult

def mapFile        = node.mindMap.file
def mapFileUri     = mapFile.toURI()
def mapFileUriText = mapFileUri.toString()
def mapFileName    = mapFileUriText.drop(mapFileUriText.lastIndexOf('/') + 1)
def mapName        = mapFileName.take(mapFileName.lastIndexOf('.'))
def folderName     = mapName + '_files'
def folderUri      = (mapFile.parentFile.toURI().toString() + folderName +'/').toURI()

def folder         = new File(folderUri)
println "1. Folder to evaluate:    $folder"
println "   a. folder '$folderName'  does ${folder.exists()?'': 'not '}exist"
println "   b. folder '$folderName'  is ${folder.isDirectory()?'': 'not '}a directory\n"

if( !(folder.exists() && folder.isDirectory() )) return 'ABORTED: ---- folder not encountered ---- '

def listOfFiles = []
def sortByTypeThenName = { a, b -> a.isFile() != b.isFile() ? a.isFile() <=> b.isFile() : a.name <=> b.name }
folder.traverse(
    type         : FileType.FILES,
    sort         : sortByTypeThenName
){it ->
    listOfFiles << mapFileUri.resolve(it.toURI())
}
println "    i) ${listOfFiles.size()} file(s) found in '$folderName' folder"

def imageNodes = node.mindMap.root.find{n -> n.externalObject.uri? true : false }
def listOfImagesInMap = imageNodes.collect{ mapFileUri.resolve(it.externalObject.uri)}.unique()
println "   ii) ${listOfImagesInMap.size()} linked image(s) found in '$mapName' mind map"

def fileNodes = node.mindMap.root.find{n -> n.link && n.link.uri && n.link.file }
def listOfFilesInMap = fileNodes.collect{ mapFileUri.resolve(it.link.uri)}.unique()
println "  iii) ${listOfFilesInMap.size()} linked file(s) found in '$mapName' mind map\n"

def filesInFolderAndNotInMap = (listOfFiles - listOfImagesInMap - listOfFilesInMap)

def q = filesInFolderAndNotInMap.size()
println "  -->  $q file${q==1 ? '' : 's'} present in '$folderName' folder that ${q==1 ? 'is' : 'are'} not linked in '$mapName' mind map\n"

filesInFolderAndNotInMap.eachWithIndex{ uri, i ->
    def file        = new File (uri)
    def filePath    = file.path
    def fileDeleted = file.delete()
    def msg         = "${i+1}. file${fileDeleted ? '' : ' NOT'} deleted: $filePath"
    println msg
}
return 'done'

```

hope it helps,

edo

**Reply by theelectricsoul** [reply]
Great. thank you.....    it is definitely a piece worth playing with.

**Comment by EdoFro** [comment]
Here are also other scripts I tested a time ago. They are for Windows.

1. You have to select the node with the link to the file
2. execute any of the following scripts
3. it will open the Files Explorer with the file selected

I don't remember what difference they had one to another, or when to use one or the other. Both work.

### script using Runtime.getRuntime().exec
```groovy
import java.io.File;
import java.io.IOException;

File f = node.link.file?:(new File(System.getenv("programfiles")))

Process proc

try {
    def comando = "explorer.exe  /select," + f.getAbsolutePath()
    proc = Runtime.getRuntime().exec(comando);
} catch (Exception ex) {
    System.out.println("Error - " + ex);
}

proc.waitFor()
```

### script using ProcessBuilder
```groovy
import java.io.IOException;

File f = node.link.file?:(new File(System.getenv("programfiles")))

Process proc

try {
    def comando = "explorer.exe  /select," + f.getAbsolutePath()
    pBuilder = new ProcessBuilder(comando.split(" "));
    proc = pBuilder.start();
} catch (Exception ex) {
    System.out.println("Error - " + ex);
}

proc.waitFor()
```

hope they help,

edo

**Reply by theelectricsoul** [reply]
also nice, because now drag and drop into other applications is possible :)

**Reply by theelectricsoul** [reply]
It will return a path looking like this :) I'll play around with it to make it work. :) 

![image](https://github.com/freeplane/freeplane/assets/10260832/12b0c6dc-429c-4144-8c46-b38b5289fdb6)

**Reply by EdoFro** [reply]
@theelectricsoul, @macmarrum,
Lately I have realized that it is better to use "Uri" than "file" from the nodes links.

For example, in the following script, we get to the same node file, but by resolving its uri against the map, we don't get tangled up if the link is absolute or relative or if the file is on another disk than the map or not.

```groovy
def nodeFile = node.link.file
println "nodeFile  :   $nodeFile"

def nodeUri = node.link.uri
println "nodeUri   :   $nodeUri"

def mapUri  = node.mindMap.file.toURI()
println "mapUri    :   $mapUri"

def nodeUriCorrected = mapUri.resolve(nodeUri)
println "nodeUriCorrected  :   $nodeUriCorrected"

def nodeFileCorrected = new File(nodeUriCorrected)
println "nodeFileCorrected :   $nodeFileCorrected"


println "nodeFileCorrected == nodeFile   :   ${nodeFileCorrected == nodeFile}"
println "nodeFile.exists()               :   ${nodeFile.exists()}" 
println "nodeFileCorrected.exists()      :   ${nodeFileCorrected.exists()}"

```

please make this three test:
1. file with absolute link (same or other drive as mindmap)
2. file with relative link but in the same drive as the mindmap
3. file with relative link but in another drive

I get the folowing result for:
#### case 1
```
nodeFileCorrected == nodeFile   :   true
nodeFile.exists()               :   true
nodeFileCorrected.exists()      :   true
```

#### case 2
```
nodeFileCorrected == nodeFile   :   false
nodeFile.exists()               :   true
nodeFileCorrected.exists()      :   true
```

#### case 3
```
nodeFileCorrected == nodeFile   :   false
nodeFile.exists()               :   false
nodeFileCorrected.exists()      :   true
```
ehh!?!?!?! it exists or not, but not both!!! (Schrödinger?)

I don't know why, but if the file is in another drive and/or its link is relative, then we don't have full access to it or its methods by using node.link.file

hope this helps,

edo

**Reply by macmarrum** [reply]
Hi @EdoFro 

In the source code, I can see that SetLinkByFileChooserAction, which is used for `Insert->Link->Add hyperlink (file)…`, saves a relative URI if so configured in Preferences. But when it is retrieved, `node.link.file` always converts this relative URI to an "absolute" file by adding `mindMap.file` and `link.file` together, whereas `node.link.uri` returns URI as is (relative if it was set as relative).

As you have showed, `node.link.file` provides unexpected value for case 3. You have found a bug :clap: 

I would expect that `node.link.file` returns relative file if such was saved, i.e. doesn't try to make it absolute.

**Reply by EdoFro** [reply]
@theelectricsoul,

> It will return a path looking like this :) I'll play around with it to make it work. :)

Try `f.getCannonicalPath()`. I think that's the way to get a "real" absolute path from a file
